third-party/starpu/starpu-1.0.0rc4/tools/starpu_perfmodel_display.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.6.
.TH STARPU_PERFMODEL_DISPLAY "1" "March 2012" "starpu_perfmodel_display 1.0" "User Commands"
.SH NAME
starpu_perfmodel_display \- manual page for starpu_perfmodel_display 1.0
.SH SYNOPSIS
.B starpu_perfmodel_display
[ \fIoptions \fR]
.SH DESCRIPTION
Display a given perfmodel
.PP
One must specify either \fB\-l\fR or \fB\-s\fR
Options:
.TP
\fB\-l\fR
display all available models
.TP
\fB\-s\fR <symbol>
specify the symbol
.TP
\fB\-p\fR <parameter>
specify the parameter (e.g. a, b, c, mean, stddev)
.TP
\fB\-a\fR <arch>
specify the architecture (e.g. cpu, cpu:k, cuda, gordon)
.TP
\fB\-f\fR <footprint>
display the history\-based model for the specified footprint
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-v\fR, \fB\-\-version\fR
output version information and exit
.PP
Reports bugs to <starpu\-devel@lists.gforge.inria.fr>.


CHANGES
====================
Release Changes List
====================

=============
version 1.4.0
=============

Seventh public release of Chapel, October 20, 2011

Highlights (see below for details)
----------------------------------
- added 'extern', 'export', and 'inline' keywords
- new optimizations for whole-array assignments and remote array accesses
- improved support for Qthreads and Nanos++ tasking, including multi-locale
- added dynamic range iterators via an 'AdvancedIters' module
- continued improving range semantics and implementation
- added support for # operator on dense rectangular domains and arrays
- made associative domain index modifications parallel-safe
- added --static/--dynamic linking flags
- improved communication diagnostics capabilities
- new example codes: SSCA2; primers on sync vars, leader-follower iterators
- updated documentation
- many bug fixes

Syntactic/Naming Changes
------------------------
- renamed '_extern' keyword to 'extern'
- renamed 'maxThreadsPerLocale' to 'numThreadsPerLocale'
  (see doc/README.tasks and doc/README.executing for new definitions)
- made it illegal for a string literal to contain an unescaped newline
- renamed dsiIndexLocale to dsiIndexToLocale in domain map standard interface
  (see doc/technotes/README.dsi)
- reserved 'ref' as a keyword for future use
- renamed 'iterator' enumeration to 'iterKind' for leader-follower iterators
- renamed 'follower' argument to follower iterators 'followThis'

Semantic Changes / Changes to Chapel Language
---------------------------------------------
- added 'export' keyword to support calling Chapel functions from C
- added 'inline' keyword to force a function to be inlined
- added support to # operator for dense rectangular domains and arrays
- removed support for tensor iteration from the language definition
- removed the realm concept from the language
- ambiguous ranges are now == if their representations match
- disabled promotion of the '..' operator

Newly Implemented Features
--------------------------
- improved qthreads and nanox task layers to support multilocale execution
  (see doc/README.tasks)
- adding and removing indices from associative domains is now parallel-safe
- updated the communication diagnostics interface
  (see doc/technotes/README.comm-diagnostics)

Standard Modules
----------------
- added an 'AdvancedIters' module with various dynamic scheduling algorithms
- added floor/ceiling/mod functions for integers to 'Math' module
- added new sorting routines to standard 'Sort' module
- added a 'UtilMath' module with other integer ceiling/floor functions
- added a param version of abs() for integers to 'Math' module

Documentation
-------------
- spec updates:
  - added a new chapter on 'Interoperability'
  - refreshed 'Data Parallelism', 'Task Parallelism', and 'Conversions' chapters
  - updated 'Ranges' chapter significantly
  - minor updates to 'Domains', 'Arrays', and 'Classes' chapters
  - split 'Memory Consistency Model' into its own chapter (still poorly defined)
  - documented new Advanced Iterators and Sort routines in 'Standard Modules'
- added documentation about the compiler's module search path to man page/technotes
  (see 'man chpl' and doc/technotes/README.module_search)
- added note about parallel safety to doc/technotes/README.extern
- generally refreshed README* and other text-based documentation files

Example Codes
-------------
- added a(n in-progress) implementation of SSCA2 to the release examples
  (see examples/ssca2)
- added a primer for sync and single variables
  (see examples/primers/syncsingle.chpl)
- added a primer on leader-follower iterators (user-defined forall scheduling)
  (see examples/primers/leaderfollower.chpl)
- added an example of creating a private locale view to distributions primer
  (see examples/primers/distributions.chpl)
- improved the ranges primer
  (see examples/primers/ranges.chpl)
- added a description of subdomains to the domains primer
  (see examples/primers/domains.chpl)
- minor wordsmithing for the procedures primer
  (see examples/primers/procedures.chpl)

Platform-specific Notes
-----------------------
- improved Makefile settings for IBM/AIX platforms and compilers

Feature Improvements
--------------------
- several improvements to the implementation of ranges
  (see 'Ranges' chapter of spec for details)
- made casts-to-string produce the same string as write() does

Compiler Flags
--------------
- added --static/--dynamic flags to specify how the binary should be linked

Interoperability Changes
------------------------
- made 'extern' a keyword; '_extern' is deprecated

Error Message Improvements
--------------------------
- added a compiler warning for param functions with declared return types
- improved a few other difficult compiler error messages

Bug Fixes / New Semantic Checks (for old semantics)
---------------------------------------------------
- fixed the result widths of complex math operations
- improved parsing of escaped characters in string literals
- deleting a record now generates an error
- added a variety of bug fixes for range operations
- fixed a bug with passing strings to external functions for --no-local cases
- fixed a bug for var functions that returned nothing
- fixed a bug related to terminating a CHPL_COMM=gasnet program early
- fixed a bug in which ref temps were not inserted for virtual method calls
- fixed a bug related to removing a label from the AST but not its goto
- fixed bug with applying serial statement to optimized coforall+on loops
- fixed bugs regarding single variables not being treated like sync
- fixed a bug related to incorrect complaints about ambiguous module paths
- fixed a bug related to followers accepting empty ranges
- fixed a few reference-after-deallocation bugs in the compiler
- fixed a bug related to flattening var functions multiple times

Packaging Changes
-----------------
- removed the second-level 'chapel/' directory that releases used to have

third-party/starpu/starpu-1.0.0rc3/tools/starpu_perfmodel_display.c
/* StarPU --- Runtime system for heterogeneous multicore architectures.
 *
 * Copyright (C) 2011  Université de Bordeaux 1
 * Copyright (C) 2011  Centre National de la Recherche Scientifique
 * Copyright (C) 2011  Télécom-SudParis
 *
 * StarPU is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or (at
 * your option) any later version.
 *
 * StarPU is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * See the GNU Lesser General Public License in COPYING.LGPL for more details.
 */

#include <config.h>
#include <assert.h>
#include <getopt.h>
#include <unistd.h>
#include <stdio.h>

#include <starpu.h>
#include <starpu_perfmodel.h>

#ifdef __MINGW32__
#include <windows.h>
#endif

#define PROGNAME "starpu_perfmodel_display"

static struct starpu_perfmodel model;

/* display all available models */
static int list = 0;
/* what kernel ? */
static char *symbol = NULL;
/* what parameter should be displayed ? (NULL = all) */
static char *parameter = NULL;
/* which architecture ? (NULL = all)*/
static char *arch = NULL;
/* should we display a specific footprint ? */
unsigned display_specific_footprint;
uint32_t specific_footprint;

static void usage(char **argv)
{
	fprintf(stderr, "Display a given perfmodel\n\n");
	fprintf(stderr, "Usage: %s [ options ]\n", PROGNAME);
        fprintf(stderr, "\n");
        fprintf(stderr, "One must specify either -l or -s\n");
        fprintf(stderr, "Options:\n");
        fprintf(stderr, "   -l                  display all available models\n");
        fprintf(stderr, "   -s <symbol>         specify the symbol\n");
        fprintf(stderr, "   -p <parameter>      specify the parameter (e.g. a, b, c, mean, stddev)\n");
        fprintf(stderr, "   -a <arch>           specify the architecture (e.g. cpu, cpu:k, cuda, gordon)\n");
	fprintf(stderr, "   -f <footprint>      display the history-based model for the specified footprint\n");
	fprintf(stderr, "   -h, --help          display this help and exit\n");
	fprintf(stderr, "   -v, --version       output version information and exit\n\n");
        fprintf(stderr, "Reports bugs to <"PACKAGE_BUGREPORT">.");
}

static void parse_args(int argc, char **argv)
{
	int c;

	static struct option long_options[] =
	{
		{"arch",      required_argument, NULL, 'a'},
		{"footprint", required_argument, NULL, 'f'},
		{"help",      no_argument,       NULL, 'h'},
		/* XXX Would be cleaner to set a flag */
		{"list",      no_argument,       NULL, 'l'},
		{"parameter", required_argument, NULL, 'p'},
		{"symbol",    required_argument, NULL, 's'},
		{"version",   no_argument,       NULL, 'v'},
		{0, 0, 0, 0}
	};

	int option_index;
	while ((c = getopt_long(argc, argv, "ls:p:a:f:h", long_options, &option_index)) != -1) {
	switch (c) {
                case 'l':
                        /* list all models */
                        list = 1;
                        break;

		case 's':
			/* symbol */
			symbol = optarg;
			break;

		case 'p':
			/* parameter (eg. a, b, c, mean, stddev) */
			parameter = optarg;
			break;

		case 'a':
			/* architecture (cpu, cuda, gordon) */
			arch = optarg;
			break;

		case 'f':
			/* footprint */
			display_specific_footprint = 1;
			sscanf(optarg, "%08x", &specific_footprint);
			break;

		case 'h':
			usage(argv);
			exit(EXIT_SUCCESS);

		case 'v':
			(void) fprintf(stdout, "%s %d.%d\n",
				PROGNAME, STARPU_MAJOR_VERSION,
				STARPU_MINOR_VERSION);
			exit(EXIT_SUCCESS);

		case '?':
		default:
			fprintf(stderr, "Unrecognized option: -%c\n", optopt);
	}
	}

	if (!symbol && !list)
	{
		fprintf(stderr, "Incorrect usage, aborting\n");
                usage(argv);
		exit(-1);
	}
}

static void display_history_based_perf_model(struct starpu_per_arch_perfmodel *per_arch_model)
{
	struct starpu_history_list *ptr;

	ptr = per_arch_model->list;

	if (!parameter && ptr)
		fprintf(stderr, "# hash\t\tsize\t\tmean\t\tdev\t\tn\n");

	while (ptr) {
		struct starpu_history_entry *entry = ptr->entry;
		if (!display_specific_footprint || (entry->footprint == specific_footprint))
		{
			if (!parameter)
			{	
				/* There isn't a parameter that is explicitely requested, so we display all parameters */
				printf("%08x\t%-15lu\t%-15le\t%-15le\t%u\n", entry->footprint,
					(unsigned long) entry->size, entry->mean, entry->deviation, entry->nsample);
			}
			else {
				/* only display the parameter that was specifically requested */
				if (strcmp(parameter, "mean") == 0) {
					printf("%-15le\n", entry->mean);
				}
		
				if (strcmp(parameter, "stddev") == 0) {
					printf("%-15le\n", entry->deviation);
					return;
				}
			}
		}

		ptr = ptr->next;
	}
}

static void display_perf_model(struct starpu_perfmodel *model, enum starpu_perf_archtype arch, unsigned nimpl)
{
	struct starpu_per_arch_perfmodel *arch_model = &model->per_arch[arch][nimpl];
	char archname[32];

	if (arch_model->regression.nsample || arch_model->regression.valid || arch_model->regression.nl_valid || arch_model->list) {

		starpu_perfmodel_get_arch_name(arch, archname, 32, nimpl);
		fprintf(stderr, "performance model for %s\n", archname);
	}

	if (parameter == NULL)
	{
		/* no specific parameter was requested, so we display everything */
		if (arch_model->regression.nsample)
			fprintf(stderr, "\tRegression : #sample = %d\n",
				arch_model->regression.nsample);

		/* Only display the regression model if we could actually build a model */
		if (arch_model->regression.valid)
		{
			fprintf(stderr, "\tLinear: y = alpha size ^ beta\n");
			fprintf(stderr, "\t\talpha = %e\n", arch_model->regression.alpha);
			fprintf(stderr, "\t\tbeta = %e\n", arch_model->regression.beta);
		}
		else {
			//fprintf(stderr, "\tLinear model is INVALID\n");
		}
	
		if (arch_model->regression.nl_valid)
		{
			fprintf(stderr, "\tNon-Linear: y = a size ^b + c\n");
			fprintf(stderr, "\t\ta = %e\n", arch_model->regression.a);
			fprintf(stderr, "\t\tb = %e\n", arch_model->regression.b);
			fprintf(stderr, "\t\tc = %e\n", arch_model->regression.c);
		}
		else {
			//fprintf(stderr, "\tNon-Linear model is INVALID\n");
		}

		display_history_based_perf_model(arch_model);

#if 0
		char debugname[1024];
		starpu_perfmodel_debugfilepath(model, arch, debugname, 1024, nimpl);
		printf("\t debug file path : %s\n", debugname);
#endif
	}
	else {
		/* only display the parameter that was specifically requested */
		if (strcmp(parameter, "a") == 0) {
			printf("%e\n", arch_model->regression.a);
			return;
		}

		if (strcmp(parameter, "b") == 0) {
			printf("%e\n", arch_model->regression.b);
			return;
		}

		if (strcmp(parameter, "c") == 0) {
			printf("%e\n", arch_model->regression.c);
			return;
		}

		if (strcmp(parameter, "alpha") == 0) {
			printf("%e\n", arch_model->regression.alpha);
			return;
		}

		if (strcmp(parameter, "beta") == 0) {
			printf("%e\n", arch_model->regression.beta);
			return;
		}

		if (strcmp(parameter, "path-file-debug") == 0) {
			char debugname[256];
			starpu_perfmodel_debugfilepath(model, arch, debugname, 1024, nimpl);
			printf("%s\n", debugname);
			return;
		}

		if ((strcmp(parameter, "mean") == 0) || (strcmp(parameter, "stddev"))) {
			display_history_based_perf_model(arch_model);
			return;
		}

		/* TODO display if it's valid ? */

		fprintf(stderr, "Unknown parameter requested, aborting.\n");
		exit(-1);
	}
}

static void display_all_perf_models(struct starpu_perfmodel *model)
{
	if (arch == NULL)
	{
		/* display all architectures */
		unsigned archid;
		unsigned implid;
		for (archid = 0; archid < STARPU_NARCH_VARIATIONS; archid++) {
			for (implid = 0; implid < STARPU_MAXIMPLEMENTATIONS; implid++) { /* Display all codelets on each arch */
				display_perf_model(model, (enum starpu_perf_archtype) archid, implid);
			}
		}
	}
	else {
		if (strcmp(arch, "cpu") == 0) {
			unsigned implid;
			for (implid = 0; implid < STARPU_MAXIMPLEMENTATIONS; implid++)
				display_perf_model(model, STARPU_CPU_DEFAULT,implid); /* Display all codelets on cpu */
			return;
		}

		int k;
		if (sscanf(arch, "cpu:%d", &k) == 1)
		{
			/* For combined CPU workers */
			if ((k < 1) || (k > STARPU_MAXCPUS))
			{
				fprintf(stderr, "Invalid CPU size\n");
				exit(-1);
			}

			unsigned implid;
			for (implid = 0; implid < STARPU_MAXIMPLEMENTATIONS; implid++)
				display_perf_model(model, (enum starpu_perf_archtype) (STARPU_CPU_DEFAULT + k - 1), implid);
			return;
		}

		if (strcmp(arch, "cuda") == 0) {
			unsigned archid;
			unsigned implid;
			for (archid = STARPU_CUDA_DEFAULT; archid < STARPU_CUDA_DEFAULT + STARPU_MAXCUDADEVS; archid++) {
				for (implid = 0; implid <STARPU_MAXIMPLEMENTATIONS; implid ++) {
					char archname[32];
					starpu_perfmodel_get_arch_name((enum starpu_perf_archtype) archid, archname, 32, implid);
					fprintf(stderr, "performance model for %s\n", archname);
					display_perf_model(model, (enum starpu_perf_archtype) archid, implid);
				}
			}
			return;
		}

		/* There must be a cleaner way ! */
		int gpuid;
		int nmatched;
		nmatched = sscanf(arch, "cuda_%d", &gpuid);
		if (nmatched == 1)
		{
			unsigned archid = STARPU_CUDA_DEFAULT+ gpuid;
			unsigned implid;
			for (implid = 0; implid < STARPU_MAXIMPLEMENTATIONS; implid++)
				display_perf_model(model, (enum starpu_perf_archtype) archid, implid);
			return;
		}

		if (strcmp(arch, "gordon") == 0) {
			fprintf(stderr, "performance model for gordon\n");
			unsigned implid;
			for (implid = 0; implid < STARPU_MAXIMPLEMENTATIONS; implid++)
				display_perf_model(model, STARPU_GORDON_DEFAULT, implid);
			return;
		}

		fprintf(stderr, "Unknown architecture requested, aborting.\n");
		exit(-1);
	}
}

int main(int argc, char **argv)
{
//	assert(argc == 2);

#ifdef __MINGW32__
	WSADATA wsadata;
	WSAStartup(MAKEWORD(1,0), &wsadata);
#endif

	parse_args(argc, argv);

        if (list) {
                int ret = starpu_list_models(stdout);
                if (ret) {
                        fprintf(stderr, "The performance model directory is invalid\n");
                        return 1;
                }
        }
        else {
                int ret = starpu_load_history_debug(symbol, &model);
                if (ret == 1)
                        {
                                fprintf(stderr, "The performance model could not be loaded\n");
                                return 1;
                        }

                display_all_perf_models(&model);
        }

	return 0;
}
